src/lib/receipt_chain_database_lib/database.ml

```diff
@@ -0,0 +1,253 @@
+open Core_kernel
+
+type ('key, 'value) tree_node =
+  | Root of {children: 'key list}
+  | Child of {parent: 'key; children: 'key list; value: 'value}
+[@@deriving sexp]
+
+module Make
+    (Transaction : Intf.Transaction)
+    (Receipt_chain_hash : Intf.Receipt_chain_hash
+                          with type transaction_payload := Transaction.payload)
+    (Key_value_db : Key_value_database.S
+                    with type key := Receipt_chain_hash.t
+                     and type value :=
+                                (Receipt_chain_hash.t, Transaction.t) tree_node) :
+  Intf.Test.S
+  with type receipt_chain_hash := Receipt_chain_hash.t
+   and type transaction := Transaction.t
+   and type database := Key_value_db.t =
+struct
+  type t = Key_value_db.t
+
+  let create ~directory = Key_value_db.create ~directory
+
+  (* Prove will provide  a merkle list of a proving receipt h_1 
+  and it's corresponding transaction t_1 to  a resulting_receipt h_k 
+  and it's corresponding transaction t_k, inclusively *)
+  let prove t ~proving_receipt ~resulting_receipt =
```

Comment:
It's more efficient to walk backwards than doing a DFS, right? Since you don't have to explore all branches

Reply:
Walk backwards from the resulting receipt to the proving_receipt